### PR TITLE
Implementation of VehicleData

### PR DESCRIPTION
**Pull request by hsyyid**
Implementation of `VehicleData`.
- [x] Registration of field getters and setters
- [x] Accurately used the appropriate `AbstractData` implementation

Implementation of `REPLACE_WITH_IMMUTABLE_MANIPULATOR`
- [x] Accurately extended the appropriate `AbstractImmutableData` implementation
- [x] Accurately instantiating final instance fields with an `ImmutableValue` counter part for any value getters
- [x] If necessary, creating a new `ImmutableValue` for a value that should not be cached, or, using the existing caching utils.

Implementation of the `DATA_MANIPULATOR_BUILDER`
- [x] In the `build()` method, checked if the container contained the keys required, if it doesn't, return `Optional.empty()`

Implementation of the `VALUE_PROCESSOR`(s):
- [x] Appropriately extended `AbstractSpongeValueProcessor`
- [x] Individual processors for each source type possible (one for `ItemStack`/`TileEntity`/`Entity` as necessary).

Registration:
- [x] Registered the `Key` correctly by `FIELD_NAME` in the `KeyRegistry`
- [x] Registered the `DataProcessor`s and `DataManipulatorBuilder` in `SpongeSerializationRegistry`
- [x] Registered the `ValueProcessor`s in the `SpongeSerializationRegistry`

Test Plugin code:

```java
package io.github.hsyyid.testplugin.cmds;

import org.spongepowered.api.data.DataTransactionResult;
import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.data.manipulator.mutable.entity.VehicleData;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.text.Texts;
import org.spongepowered.api.text.format.TextColors;
import org.spongepowered.api.util.command.CommandException;
import org.spongepowered.api.util.command.CommandResult;
import org.spongepowered.api.util.command.CommandSource;
import org.spongepowered.api.util.command.args.CommandContext;
import org.spongepowered.api.util.command.source.CommandBlockSource;
import org.spongepowered.api.util.command.source.ConsoleSource;
import org.spongepowered.api.util.command.spec.CommandExecutor;

import java.util.Optional;

public class VehicleDataCommandExecutor implements CommandExecutor
{
	public CommandResult execute(CommandSource src, CommandContext ctx) throws CommandException
	{
		if (src instanceof Player)
		{
			Player player = (Player) src;

			Optional<VehicleData> optionalData = player.getOrCreate(VehicleData.class);

			if (optionalData.isPresent())
			{
				VehicleData data = optionalData.get().fill(player).get();

				if (data.get(Keys.VEHICLE).isPresent())
				{
					DataTransactionResult result = player.remove(VehicleData.class);
					System.out.println(result.getType().toString());
				}
				else
				{
					System.out.println("Vehicle not found");
				}

				if (data.get(Keys.BASE_VEHICLE).isPresent())
				{
					System.out.println("Player " + player.getName() + " 's base vehicle is " + data.get(Keys.BASE_VEHICLE).get().getType());
				}
				else
				{
					System.out.println("Base vehicle not found");
				}
			}
			else
			{
				System.out.println("Error! VehicleData not present!");
			}
		}
		else if (src instanceof ConsoleSource)
		{
			src.sendMessage(Texts.of(TextColors.DARK_RED, "Error! ", TextColors.RED, "Must be an in-game player to use /vehicle!"));
		}
		else if (src instanceof CommandBlockSource)
		{
			src.sendMessage(Texts.of(TextColors.DARK_RED, "Error! ", TextColors.RED, "Must be an in-game player to use /vehicle!"));
		}

		return CommandResult.success();
	}
}

```

### CODE REVIEW

**Comment by JBYoshi** [comment]
You can just push to a branch linked to an already open PR to update it.

File: src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java [hunk 1]
```diff
@@ -105,7 +105,7 @@ public void setMaxHealth(double maxHealth) {
     }
 
     public void damageD(double amount) {
-        Living thisEntity = (Living) this;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java]: Unnecessary.

File: src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java [hunk 2]
```diff
@@ -125,7 +125,7 @@ public double getHealthD() {
     }
 
     public void setHealthD(double health) {
-        Living thisEntity = (Living) this;
+        Living thisEntity = this;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/living/MixinEntityLivingBase.java]: Unnecessary.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    public ImmutableSpongePassengerData(Entity vehicle, Entity baseVehicle) {
+        super(ImmutablePassengerData.class);
+        this.vehicle = vehicle;
+        this.baseVehicle = baseVehicle;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableBoundedValue<Entity> vehicle() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.VEHICLE, this.vehicle, this.vehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java]: DO NOT USE CACHING ON ENTITIES!

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return new MemoryDataContainer().set(Keys.VEHICLE.getQuery(), this.vehicle).set(Keys.BASE_VEHICLE.getQuery(), this.baseVehicle);
+    }
+
+    @Override
+    public int compareTo(ImmutablePassengerData o) {
+        return ComparisonChain.start().compare(o.vehicle().get().getType().getName(), this.vehicle.getType().getName())
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java]: The names really don't matter, the `UUID` checks do. Just use those.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.entity.ImmutablePassengerData;
+import org.spongepowered.api.data.manipulator.mutable.entity.PassengerData;
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.DataProcessor;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongePassengerData;
+
+public class PassengerDataProcessor implements DataProcessor<PassengerData, ImmutablePassengerData> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: Extend `AbstractSpongeDataProcessor`. Seriously.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.ValueProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+public class BaseVehicleValueProcessor implements ValueProcessor<Entity, Value<Entity>> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java]: Extend AbstractSpongeValueProcessor.

File: SpongeAPI [hunk 7]
```diff
@@ -1 +1 @@
-Subproject commit 3ea22a749fc1725efedc8f036eaa9a49c5a5724a
+Subproject commit f164da66e3d0294f30379810ad9b82afa798a519
```
Comment (gabizou) [inline SpongeAPI]: Why are you updating the submodule?

**Comment by hsyyid** [comment]
@gabizou Accidentally.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return new MemoryDataContainer().set(Keys.VEHICLE.getQuery(), this.vehicle).set(Keys.BASE_VEHICLE.getQuery(), this.baseVehicle);
+    }
+
+    @Override
+    public int compareTo(ImmutablePassengerData o) {
+        return ComparisonChain.start().compare(o.vehicle().get().getUniqueId(), this.vehicle.getUniqueId())
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java]: For readability, try just 
```
ComparisonChain.start()
  .compare(...)
  .compare(...)
  .result();
```

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongePassengerData.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    public SpongePassengerData setBaseVehicle(Entity value) {
+        this.baseVehicle = checkNotNull(value);
+        return this;
+    }
+
+    @Override
+    public int compareTo(PassengerData o) {
+        return ComparisonChain.start().compare(o.vehicle().get().getUniqueId(), this.vehicle.getUniqueId())
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongePassengerData.java]: Likewise here, for readability.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongePassengerData.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @Override
+    public ImmutablePassengerData asImmutable() {
+        return new ImmutableSpongePassengerData(this.vehicle, this.baseVehicle);
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return new MemoryDataContainer().set(Keys.VEHICLE, this.vehicle).set(Keys.BASE_VEHICLE, this.baseVehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongePassengerData.java]: Likewise here, for readability.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        } else {
+            return Optional.absent();
+        }
+
+    }
+
+    @Override
+    public Optional<PassengerData> fill(DataHolder dataHolder, PassengerData manipulator, MergeFunction overlap) {
+        if (supports(dataHolder)) {
+            final PassengerData merged = overlap.merge(checkNotNull(manipulator).copy(), from(dataHolder).get());
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: `checkNotNull(overlap).merge....`

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    @Override
+    public Optional<PassengerData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final SpongePassengerData passengerData = new SpongePassengerData();
+            final Entity vehicle = (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) vehicle;
+            while (currentEntity != null && currentEntity.isRiding()) {
+                currentEntity = currentEntity.ridingEntity;
+            }
+            final Entity baseVehicle = (Entity) currentEntity;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: I don't see any checks whether the vehicle is null.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            return Optional.<PassengerData>of(new SpongePassengerData(vehicle, baseVehicle));
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public int getPriority() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: don't override this.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        return offerToStore(container, new SpongeValue<Entity>(Keys.BASE_VEHICLE, value, value));
+    }
+
+    @Override
+    public DataTransactionResult removeFrom(ValueContainer<?> container) {
+        return DataTransactionBuilder.failNoData();
+    }
+
+    @Override
+    public int getPriority() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java]: don't override this.

File: src/main/java/org/spongepowered/common/mixin/api/text/MixinText.java [hunk 15]
```diff
@@ -33,8 +33,9 @@
 import org.spongepowered.api.text.action.ClickAction;
 import org.spongepowered.api.text.action.HoverAction;
 import org.spongepowered.api.text.action.ShiftClickAction;
+import org.spongepowered.api.text.format.TextColor;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/api/text/MixinText.java]: This entire file shouldn't be touched by you.

File: src/main/java/org/spongepowered/common/mixin/core/server/MixinMinecraftServer.java [hunk 16]
```diff
@@ -555,14 +555,6 @@ public boolean unloadWorld(World world) {
     }
 
     @Override
-    public Optional<WorldProperties> getDefaultWorld() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/server/MixinMinecraftServer.java]: Nor this.

File: src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/PassengerDataBuilder.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    public Optional<PassengerData> createFrom(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            final SpongePassengerData passengerData = new SpongePassengerData();
+            final Entity vehicle = (Entity) ((net.minecraft.entity.Entity) dataHolder).ridingEntity;
+            net.minecraft.entity.Entity currentEntity = (net.minecraft.entity.Entity) vehicle;
+            while (currentEntity != null && currentEntity.isRiding()) {
+                currentEntity = currentEntity.ridingEntity;
+            }
+            final Entity baseVehicle = (Entity) currentEntity;
+            return Optional.<PassengerData>of(passengerData.setVehicle(vehicle).setBaseVehicle(baseVehicle));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/builder/manipulator/mutable/entity/PassengerDataBuilder.java]: You still aren't doing any null checks. Seriously.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public PassengerData asMutable() {
+        return new SpongePassengerData(this.vehicle, this.baseVehicle);
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return new MemoryDataContainer().set(Keys.VEHICLE.getQuery(), this.vehicle).set(Keys.BASE_VEHICLE.getQuery(), this.baseVehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java]: for readability, do the same as the comparisonchain here.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            net.minecraft.entity.Entity entity = ((net.minecraft.entity.Entity) dataHolder);
+            if (entity.isRiding()) {
+                entity.mountEntity(null);
+                return DataTransactionBuilder.builder().result(DataTransactionResult.Type.SUCCESS).build();
+
+            } else {
+                return DataTransactionBuilder.builder().result(DataTransactionResult.Type.FAILURE).build();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: This should be a success regardless.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            net.minecraft.entity.Entity entity = ((net.minecraft.entity.Entity) dataHolder);
+            if (entity.isRiding()) {
+                entity.mountEntity(null);
+                return DataTransactionBuilder.builder().result(DataTransactionResult.Type.SUCCESS).build();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/PassengerDataProcessor.java]: You should try to check if there was an entity the dataholder was riding already, if so, provide the "replaced" result in the transaction.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+    @Override
+    public Optional<Entity> getValueFromContainer(ValueContainer<?> container) {
+        if (container instanceof net.minecraft.entity.Entity) {
+            net.minecraft.entity.Entity currentEntity = ((net.minecraft.entity.Entity) container).ridingEntity;
+            while (currentEntity != null && currentEntity.isRiding()) {
+                currentEntity = currentEntity.ridingEntity;
+            }
+            final Entity baseVehicle = (Entity) currentEntity;
+            return Optional.of(baseVehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java]: fromNullable.

**Comment by hsyyid** [comment]
@gabizou How would I provide the replaced result.

**Comment by gabizou** [comment]
You're missing a few methods to override. Please fix, make sure it compiles, etc. etc. etc.

**Comment by hsyyid** [comment]
@gabizou Tada :smile:

**Comment by hsyyid** [comment]
@gabizou What now?

**Comment by gabizou** [comment]
Please update according to the recent changes for `Entity` specific `DataProcessor`s.

**Comment by hsyyid** [comment]
Okay

**Comment by hsyyid** [comment]
@gabizou Done

File: src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java [hunk 22]
```diff
@@ -395,21 +421,14 @@
 import org.spongepowered.common.world.type.SpongeWorldTypeNether;
 import org.spongepowered.common.world.type.SpongeWorldTypeOverworld;
 
-import java.awt.Color;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/registry/SpongeGameRegistry.java]: Fix your import formatting.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Entity value) {
+        final Entity actualValue = value;
+        final ImmutableValue<Entity> proposedValue = ImmutableDataCachingUtil.getValue(ImmutableSpongeValue.class, Keys.VEHICLE, actualValue, null);
+        if (container instanceof net.minecraft.entity.Entity) {
+            final DataTransactionBuilder builder = DataTransactionBuilder.builder();
+            final ImmutableValue<Entity> newVehicleValue =
+                    ImmutableDataCachingUtil.getValue(ImmutableSpongeValue.class, Keys.VEHICLE, actualValue, null);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java]: No. no caching.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public boolean supports(ValueContainer<?> container) {
+        return container instanceof net.minecraft.entity.Entity;
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Entity value) {
+        final Entity actualValue = value;
+        final ImmutableValue<Entity> proposedValue = ImmutableDataCachingUtil.getValue(ImmutableSpongeValue.class, Keys.VEHICLE, actualValue, null);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java]: No caching.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+import org.spongepowered.api.data.value.ValueContainer;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.ImmutableDataCachingUtil;
+import org.spongepowered.common.data.processor.common.AbstractSpongeValueProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+public class VehicleValueProcessor extends AbstractSpongeValueProcessor<Entity, Value<Entity>> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/VehicleValueProcessor.java]: Why don't you just use `AbstractEntitySingleDataProcessor` instead?

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeVehicleData.java [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+        registerGettersAndSetters();
+    }
+
+    public SpongeVehicleData() {
+        this(null, null);
+    }
+
+    @Override
+    public Value<Entity> vehicle() {
+        return new SpongeValue<Entity>(Keys.VEHICLE, vehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeVehicleData.java]: Use `SpongeEntityValue` here.

File: src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeVehicleData.java [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    @Override
+    public Value<Entity> baseVehicle() {
+        return new SpongeValue<Entity>(Keys.BASE_VEHICLE, baseVehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/mutable/entity/SpongeVehicleData.java]: And here.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+    @Override
+    public boolean supports(ValueContainer<?> container) {
+        return container instanceof net.minecraft.entity.Entity;
+    }
+
+    @Override
+    public DataTransactionResult offerToStore(ValueContainer<?> container, Entity value) {
+        final Entity actualValue = value;
+        final ImmutableValue<Entity> proposedValue =
+                ImmutableDataCachingUtil.getValue(ImmutableSpongeValue.class, Keys.BASE_VEHICLE, actualValue, null);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java]: No caching.

File: src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+        if (container instanceof net.minecraft.entity.Entity) {
+            final DataTransactionBuilder builder = DataTransactionBuilder.builder();
+            final ImmutableValue<Entity> newVehicleValue =
+                    ImmutableDataCachingUtil.getValue(ImmutableSpongeValue.class, Keys.BASE_VEHICLE, actualValue, null);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/value/entity/BaseVehicleValueProcessor.java]: No caching.

**Comment by hsyyid** [comment]
@gabizou How to register an `AbstractEntitySingleDataProcessor` correctly? Currently it doesn't work

**Comment by hsyyid** [comment]
@gabizou AbstractEntitySingleDataProcessor is for DataProcessors, not ValueProcessors.

**Comment by hsyyid** [comment]
@gabizou It's all updated.

**Comment by gabizou** [comment]
Give me a bit because I'm about to introduce breaking changes for all of the data PR's. It also introduces some unit testing (really basic at the moment).

**Comment by clienthax** [comment]
Helps if you register it nub ;)

File: src/main/java/org/spongepowered/common/data/key/KeyRegistry.java [hunk 30]
```diff
@@ -98,11 +99,12 @@ private static void generateKeyMap() {
         keyMap.put("remaining_air", makeSingleKey(Integer.TYPE, Value.class, of("RemainingAir")));
         keyMap.put("fire_ticks", makeSingleKey(Integer.TYPE, MutableBoundedValue.class, of("FireTicks")));
         keyMap.put("fire_damage_delay", makeSingleKey(Integer.TYPE, MutableBoundedValue.class, of("FireDamageDelay")));
-        keyMap.put("food_level", makeSingleKey(Integer.TYPE, Value.class, of("FoodLevel")));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/key/KeyRegistry.java]: Why are you removing this?

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeVehicleData.java [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+import org.spongepowered.api.data.manipulator.mutable.entity.VehicleData;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeVehicleData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class ImmutableSpongeVehicleData extends AbstractImmutableData<ImmutableVehicleData, VehicleData> implements ImmutableVehicleData {
+
+    private final Entity vehicle;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeVehicleData.java]: Should be stored as a `WeakReference<Entity>`.

File: src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeVehicleData.java [hunk 32]
```diff
@@ -0,0 +1,100 @@
...
+    public ImmutableSpongeVehicleData(Entity vehicle, Entity baseVehicle) {
+        super(ImmutableVehicleData.class);
+        this.vehicle = vehicle;
+        this.baseVehicle = baseVehicle;
+        registerGetters();
+    }
+
+    @Override
+    public ImmutableValue<Entity> vehicle() {
+        return new ImmutableSpongeValue<Entity>(Keys.VEHICLE, vehicle);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongeVehicleData.java]: Pretty sure I created [`ImmutableSpongeEntityValue`](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/data/value/immutable/common/ImmutableSpongeEntityValue.java) for this specifically.

**Comment by hsyyid** [comment]
@gabizou "Why are you removing this?" It's registered twice.

**Comment by gabizou** [comment]
> @gabizou "Why are you removing this?" It's registered twice.

That's not what I meant by commenting, I meant, comment on the commit that added it twice or make an issue.

**Comment by gabizou** [comment]
@hsyyid I'm moving this to a branch for further testing.

**Comment by hsyyid** [comment]
@gabizou Okay :)

**Pull request closed without merging**